File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <bits/stdc++.h>
#define MAX 1000000

using namespace std;

int n, m;
int tree[MAX];

void input() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  cin >> n >> m;

  for (int i = 0; i < n; i++) {
    cin >> tree[i];
  }
}

bool solve(long long h) {
  long long cur = 0;

  for (int i = 0; i < n; i++) {
    long long length = tree[i] - h;

    if (tree[i] - h > 0) {
      cur += tree[i] - h;
    }
  }

  return cur >= m;
}

int main() {
  input();

  long long st = 0, en = *max_element(tree, tree + n);

  while (st < en) {
    long long mid = (st + en + 1) / 2;

    if (solve(mid)) {
      st = mid;
    } else
      en = mid - 1;
  }

  cout << st;
  return 0;
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
const fs = require("fs");
const stdin = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [, m] = stdin[0].split(" ").map((num) => parseInt(num));
const trees = stdin[1].split(" ").map((num) => parseInt(num));

const solve = (h) => {
  let cur = 0;

  trees.forEach((tree) => {
    if (tree - h > 0) cur += tree - h;
  });

  return cur >= m;
};

let [st, en] = [0, Math.max(...trees)];

while (st < en) {
  let mid = parseInt((st + en + 1) / 2);

  if (solve(mid)) {
    st = mid;
  } else {
    en = mid - 1;
  }
}

console.log(st);
